File: frontend/src/types/metrics.ts
export interface CpuMetrics {
  percent_per_core: number[];
  percent_overall: number;
  freq_current: number | null;
  freq_max: number | null;
  load_avg: [number, number, number];
  core_count: number;
  thread_count: number;
}

export interface MemoryMetrics {
  ram_total: number;
  ram_used: number;
  ram_percent: number;
  swap_total: number;
  swap_used: number;
  swap_percent: number;
}

export interface DiskPartition {
  device: string;
  mountpoint: string;
  total: number;
  used: number;
  percent: number;
}

export interface DiskMetrics {
  partitions: DiskPartition[];
  read_rate: number;
  write_rate: number;
}

export interface NetworkInterface {
  name: string;
  bytes_sent_rate: number;
  bytes_recv_rate: number;
}

export interface NetworkMetrics {
  interfaces: NetworkInterface[];
  total_sent_rate: number;
  total_recv_rate: number;
  connection_count: number;
}

export interface GpuMetrics {
  name: string | null;
  utilization: number | null;
  vram_total: number | null;
  vram_used: number | null;
  temperature: number | null;
  available: boolean;
}

export interface TemperatureSensor {
  label: string;
  current: number;
  high: number | null;
  critical: number | null;
}

export interface ProcessInfo {
  pid: number;
  name: string;
  cpu_percent: number;
  memory_percent: number;
  status: string;
  username: string;
}

export interface SystemMetrics {
  timestamp: number;
  cpu: CpuMetrics;
  memory: MemoryMetrics;
  disk: DiskMetrics;
  network: NetworkMetrics;
  gpu: GpuMetrics;
  temperatures: TemperatureSensor[];
  processes: ProcessInfo[];
}

export interface SystemInfo {
  hostname: string;
  os: string;
  kernel: string;
  cpu_model: string;
  cpu_cores: number;
  cpu_threads: number;
  ram_total: number;
  uptime: number;
}

File: backend/app/models/metrics.py
from pydantic import BaseModel


class CpuMetrics(BaseModel):
    percent_per_core: list[float]
    percent_overall: float
    freq_current: float | None = None
    freq_max: float | None = None
    load_avg: tuple[float, float, float]
    core_count: int
    thread_count: int


class MemoryMetrics(BaseModel):
    ram_total: int
    ram_used: int
    ram_percent: float
    swap_total: int
    swap_used: int
    swap_percent: float


class DiskPartition(BaseModel):
    device: str
    mountpoint: str
    total: int
    used: int
    percent: float


class DiskMetrics(BaseModel):
    partitions: list[DiskPartition]
    read_rate: float  # bytes/s
    write_rate: float  # bytes/s


class NetworkInterface(BaseModel):
    name: str
    bytes_sent_rate: float  # bytes/s
    bytes_recv_rate: float  # bytes/s


class NetworkMetrics(BaseModel):
    interfaces: list[NetworkInterface]
    total_sent_rate: float
    total_recv_rate: float
    connection_count: int


class GpuMetrics(BaseModel):
    name: str | None = None
    utilization: float | None = None
    vram_total: int | None = None
    vram_used: int | None = None
    temperature: float | None = None
    available: bool = False


class TemperatureSensor(BaseModel):
    label: str
    current: float
    high: float | None = None
    critical: float | None = None


class ProcessInfo(BaseModel):
    pid: int
    name: str
    cpu_percent: float
    memory_percent: float
    status: str
    username: str


class SystemMetrics(BaseModel):
    timestamp: float
    cpu: CpuMetrics
    memory: MemoryMetrics
    disk: DiskMetrics
    network: NetworkMetrics
    gpu: GpuMetrics
    temperatures: list[TemperatureSensor]
    processes: list[ProcessInfo]


class SystemInfo(BaseModel):
    hostname: str
    os: str
    kernel: str
    cpu_model: str
    cpu_cores: int
    cpu_threads: int
    ram_total: int
    uptime: float
